### PR TITLE
3.1 Queue Custom Implementation

### PR DESCRIPTION
**Pull request by stoyankisyov**
1. I was not sure if the Dequeue method is supposed just to remove the head element /as explained in the task/, or remove and return it /as the build-in queue works/, so I provided you with two methods: Drop() - just removes, Dequeue removes and returns the element. 
2. Implemented a few extra methods needed for the solution of the task in the interface since we are working with abstraction and cannot access properties and class methods directly, hope it is not a problem:
    -> Count() - returning the number of elements in the queue
    -> Clone() - for creating a clone of the queue used in the Tail() method, so this way we will work with a clone and won't affect the original queue, provided as parameter. Since the only way to extract a value out of queue is by Dequeue() /not mentioning Peek() since it does not work in this case/ and that would affect the original queue /reference type/

### CODE REVIEW

File: QueueCustomImplementation/Queue.cs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+        public T Dequeue()
+        {
+            var firstItem = _elements[0];
+
+            Drop();
+
+            return firstItem;
+        }
+
```
Comment (IgorKheidorov) [inline QueueCustomImplementation/Queue.cs:32]: Drop has to be private not to provide the possibility to call it directly

File: QueueCustomImplementation/Queue.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        public void Drop()
+        {
+            if (IsEmpty())
+            {
+                throw new InvalidOperationException("Queue is empty.");
+            }
+
+            var newElements = new T[_elementCount - 1];
```
Comment (IgorKheidorov) [inline QueueCustomImplementation/Queue.cs]: Think how to implement functionality not creating new array to improve performance (hint- use array  _maxCapacity *2 and change the pointer to first element not copying elements)

File: QueueCustomImplementation/Interfaces/IGenericQueue.cs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+﻿namespace QueueCustomImplementation.Interfaces
+{
+    public interface IGenericQueue<T>
+    {
+        IGenericQueue<T> Clone();
+        T Dequeue();
```
Comment (IgorKheidorov) [inline QueueCustomImplementation/Interfaces/IGenericQueue.cs:6]: Drop is not necessary

**Review by IgorKheidorov — CHANGES_REQUESTED** [review]
Please see comments and improve performance )

File: QueueCustomImplementation/Interfaces/IGenericQueue.cs (hunk 3 above)
Comment (stoyankisyov) [reply QueueCustomImplementation/Interfaces/IGenericQueue.cs:6]: Fixed

File: QueueCustomImplementation/Queue.cs (hunk 1 above)
Comment (stoyankisyov) [reply QueueCustomImplementation/Queue.cs:32]: Fixed

File: QueueCustomImplementation/Queue.cs (hunk 2 above)
Comment (stoyankisyov) [reply QueueCustomImplementation/Queue.cs]: Fixed. This implementation with the *2 size was the best i could come up with, but it still seems unnecessarily complicated /with these checks for which side of the array the elements currently are, in every method/, so I have provided you with another file called 'AlternativeQueue' that offers a simplified solution, just by moving elements with 1 index to the left upon dequeue. Unit tests provided as well and all pass. After your review will delete worse solution

**Review by IgorKheidorov — APPROVED** [review]
Cool

**Pull request merged**